redis.go

```diff
@@ -0,0 +1,149 @@
+)
+
+const (
+	operationRedis = "Redis-"
+	logCmdName     = "command"
+	logCmdArgs     = "args"
+	logCmdResult   = "result"
+)
+
+type contextKey int
+
+const (
+	cmdStart contextKey = iota
+)
+
+// jaegerHook is go-redis jaeger hook
+type jaegerHook struct {
+	tracer opentracing.Tracer
+}
+
+// NewJaegerHook return jaegerHook
+func NewJaegerHook(tracer opentracing.Tracer) *jaegerHook {
+	return &jaegerHook{
+		tracer: tracer,
+	}
+}
+
+//BeforeProcess redis before execute action do something
+func (jh *jaegerHook) BeforeProcess(ctx context.Context, cmd redis.Cmder) (context.Context, error) {
+	if jh.tracer == nil {
+		return ctx, nil
+	}
+	span, ctx := opentracing.StartSpanFromContextWithTracer(ctx, jh.tracer, operationRedis+cmd.Name())
+
+	ctx = context.WithValue(ctx, cmdStart, span)
+	return ctx, nil
+}
+
+//AfterProcess redis after execute action do something
+func (jh *jaegerHook) AfterProcess(ctx context.Context, cmd redis.Cmder) error {
+	if jh.tracer == nil {
+		return nil
+	}
+	span, ok := ctx.Value(cmdStart).(opentracing.Span)
+	if !ok {
+		return nil
+	}
+	defer span.Finish()
+
+	span.LogFields(tracerLog.String(logCmdName, cmd.Name()))
+	span.LogFields(tracerLog.Object(logCmdArgs, cmd.Args()))
+	span.LogFields(tracerLog.Object(logCmdResult, cmd.String()))
+
+	if err := cmd.Err(); isRedisError(err) {
+		span.LogFields(tracerLog.Error(err))
+		span.SetTag(string(ext.Error), true)
+	}
+
+	return nil
+}
+
+// BeforeProcessPipeline before command process handle
+func (jh *jaegerHook) BeforeProcessPipeline(ctx context.Context, cmds []redis.Cmder) (context.Context, error) {
+	if jh.tracer == nil {
+		return ctx, nil
+	}
+
+	span, ctx := opentracing.StartSpanFromContextWithTracer(ctx, jh.tracer, operationRedis+"pipeline")
+
+	ctx = context.WithValue(ctx, cmdStart, span)
+
+	return ctx, nil
+}
+
+// AfterProcessPipeline after command process handle
+func (jh *jaegerHook) AfterProcessPipeline(ctx context.Context, cmds []redis.Cmder) error {
+	if jh.tracer == nil {
+		return nil
+	}
+
+	span, ok := ctx.Value(cmdStart).(opentracing.Span)
+	if !ok {
+		return nil
+	}
+	defer span.Finish()
+
+	hasErr := false
+	for idx, cmd := range cmds {
+		if err := cmd.Err(); isRedisError(err) {
+			hasErr = true
+		}
+		span.LogFields(tracerLog.String(jh.getPipeLineLogKey(logCmdName, idx), cmd.Name()))
+		span.LogFields(tracerLog.Object(jh.getPipeLineLogKey(logCmdArgs, idx), cmd.Args()))
+		span.LogFields(tracerLog.String(jh.getPipeLineLogKey(logCmdResult, idx), cmd.String()))
+	}
+	if !hasErr {
+		return nil
+	}
+	span.SetTag(string(ext.Error), true)
+
+	return nil
+}
+
+func (jh *jaegerHook) getPipeLineLogKey(logField string, idx int) string {
+	return logField + "-" + strconv.Itoa(idx)
+}
+
+// redisError interface
+type redisError interface {
+	error
+
+	// RedisError is a no-op function but
+	// serves to distinguish types that are Redis
+	// errors from ordinary errors: a type is a
+	// Redis error if it has a RedisError method.
+	RedisError()
+}
```

Comment:
is this copied from https://github.com/go-redis/redis/blob/master/error.go#L16 ?
why not just reference it?

Reply:
In my impression, the old version is private, maybe it was modified in v8，thanks for reminding.